reframe/frontend/statistics.py

```diff
@@ -8,82 +8,88 @@ def __init__(self, tasks=[]):
         if not hasattr(tasks, '__iter__'):
             raise TypeError('expected an iterable')
 
-        # Store test cases per partition internally
-        self._tasks_bypart = {}
-        for t in tasks:
-            partition = t.check.current_partition
-            partname = partition.fullname if partition else 'None'
-            tclist = self._tasks_bypart.setdefault(partname, [])
-            tclist.append(t)
+        self._tasks = tasks
+        self._last_retry = max([t.retry_num for t in tasks])
 
     def __repr__(self):
         return debug.repr(self)
 
-    def num_failures(self, partition=None):
-        num_fails = 0
-        if partition:
-            num_fails += len([
-                t for t in self._tasks_bypart[partition] if t.failed
-            ])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_fails += len([t for t in tclist if t.failed])
-
-        return num_fails
-
-    def num_failures_stage(self, stage):
-        num_fails = 0
-        for tclist in self._tasks_bypart.values():
-            num_fails += len([t for t in tclist if t.failed_stage == stage])
-
-        return num_fails
-
-    def num_cases(self, partition=None):
-        num_cases = 0
-        if partition:
-            num_cases += len(self._tasks_bypart[partition])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_cases += len(tclist)
-
-        return num_cases
-
-    def failure_report(self):
+    # The retry_num can be indexed from the end (e.g. -1 for the last retry)
```

Comment:
I suggest keeping the interface of statistics to the bare minimum. Only the functionality needed in other places of the framework must be implemented + a function that would allow sb to iterate over the finished tasks.